pkg/controller/certificaterequests/ca/ca.go

```diff
@@ -64,59 +70,58 @@ func init() {
-		issuerOptions: ctx.IssuerOptions,
-		secretsLister: ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
-		reporter:      crutil.NewReporter(ctx.Clock, ctx.Recorder),
+		issuerOptions:     ctx.IssuerOptions,
+		secretsLister:     ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
+		reporter:          crutil.NewReporter(ctx.Clock, ctx.Recorder),
+		templateGenerator: pki.GenerateTemplateFromCertificateRequest,
 	}
 }
 
-func (c *CA) Sign(ctx context.Context, cr *v1alpha1.CertificateRequest, issuerObj v1alpha1.GenericIssuer) (*issuerpkg.IssueResponse, error) {
+func (c *CA) Sign(ctx context.Context, cr *cmapi.CertificateRequest, issuerObj cmapi.GenericIssuer) (*issuerpkg.IssueResponse, error) {
 	log := logf.FromContext(ctx, "sign")
 
 	secretName := issuerObj.GetSpec().CA.SecretName
 	resourceNamespace := c.issuerOptions.ResourceNamespace(issuerObj)
 
 	// get a copy of the CA certificate named on the Issuer
 	caCerts, caKey, err := kube.SecretTLSKeyPair(ctx, c.secretsLister, resourceNamespace, issuerObj.GetSpec().CA.SecretName)
-	if err != nil {
-		log := logf.WithRelatedResourceName(log, issuerObj.GetSpec().CA.SecretName, resourceNamespace, "Secret")
-
-		if k8sErrors.IsNotFound(err) {
-			message := fmt.Sprintf("Referenced secret %s/%s not found", resourceNamespace, secretName)
+	if k8sErrors.IsNotFound(err) {
+		message := fmt.Sprintf("Referenced secret %s/%s not found", resourceNamespace, secretName)
 
-			c.reporter.Pending(cr, err, "MissingSecret", message)
-			log.Error(err, message)
+		c.reporter.Pending(cr, err, "SecretMissing", message)
+		log.Error(err, message)
 
-			return nil, nil
-		}
+		return nil, nil
+	}
 
-		if cmerrors.IsInvalidData(err) {
-			message := fmt.Sprintf("Failed to parse signing CA keypair from secret %s/%s", resourceNamespace, secretName)
+	if cmerrors.IsInvalidData(err) {
+		message := fmt.Sprintf("Failed to parse signing CA keypair from secret %s/%s", resourceNamespace, secretName)
 
-			c.reporter.Pending(cr, err, "ErrorParsingSecret", message)
-			log.Error(err, message)
-			return nil, nil
-		}
+		c.reporter.Pending(cr, err, "SecretInvalidData", message)
+		log.Error(err, message)
+		return nil, nil
+	}
 
+	if err != nil {
 		// We are probably in a network error here so we should backoff and retry
 		message := fmt.Sprintf("Failed to get certificate key pair from secret %s/%s", resourceNamespace, secretName)
-		c.reporter.Pending(cr, err, "ErrorGettingSecret", message)
+		c.reporter.Pending(cr, err, "SecretGetError", message)
 		log.Error(err, message)
 		return nil, err
 	}
 
-	template, err := pki.GenerateTemplateFromCertificateRequest(cr)
+	template, err := c.templateGenerator(cr)
 	if err != nil {
 		message := "Error generating certificate template"
-		c.reporter.Failed(cr, err, "ErrorSigning", message)
+		c.reporter.Failed(cr, err, "SigningError", message)
 		log.Error(err, message)
 		return nil, nil
 	}
 
 	certPEM, caPEM, err := pki.SignCSRTemplate(caCerts, caKey, template)
```

Comment:
How are we validating signed certificates in tests? Is the output of SignCSRTemplate predictable if `template` is also predictable? I thought there was an element of randomness to it 😅

Reply:
Yeah, template is where the randomness happends

```
	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
	if err != nil {
		return nil, fmt.Errorf("failed to generate serial number: %s", err.Error())
	}

	certDuration := api.DefaultCertDuration(cr.Spec.Duration)

	return &x509.Certificate{
		Version:               csr.Version,
		BasicConstraintsValid: true,
		SerialNumber:          serialNumber,

```